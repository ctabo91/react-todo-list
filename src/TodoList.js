import React, {useState} from "react";
import Todo from "./Todo";
import NewTodoForm from "./NewTodoForm";

const TodoList = () => {
    const [todos, setTodos] = useState([]);
    const addTodo = todoObj => {
        setTodos(todos => [...todos, todoObj]);
    };
    const removeTodo = id => {
        setTodos(todos => todos.filter(todo => todo.id !== id));
    };
    const toggleTodo = id => {
        setTodos(todos => 
            todos.map(todo => 
                todo.id === id ? { ...todo, isComplete: !todo.isComplete } : todo
            )
        );
    };

    const todoComponents = todos.map(todo => (
        <Todo
            key={todo.id}
            id={todo.id}
            task={todo.task}
            isComplete={todo.isComplete}
            handleRemove={removeTodo}
            handleToggle={toggleTodo}
        />
    ));

    return (
        <div className="TodoList">
            <NewTodoForm createTodo={addTodo} />
            {todoComponents}
        </div>
    );
}


export default TodoList;