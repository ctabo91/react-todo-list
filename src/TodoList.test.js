import React from "react";
import { render, fireEvent } from "@testing-library/react";
import TodoList from "./TodoList";

const addTodo = (todoList, task = "write tests") => {
  const taskInput = todoList.getByLabelText("Task:");
  fireEvent.change(taskInput, { target: { value: task }});
  const submitButton = todoList.getByText("Add Todo!");
  fireEvent.click(submitButton);
}

it("renders without crashing", () => {
  render(<TodoList />);
});

it("matches snapshot", () => {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

it("can add a todo", () => {
  const list = render(<TodoList />);
  addTodo(list);

  // expect form to clear and todo to be on the page
  expect(list.getByLabelText("Task:")).toHaveValue("");
  expect(list.getByText("write tests")).toBeInTheDocument();
  expect(list.getByText("Mark as complete")).toBeInTheDocument();
  expect(list.getByText("X")).toBeInTheDocument();
});


it("can delete a todo", () => {
  const list = render(<TodoList />);
  addTodo(list);

  fireEvent.click(list.getByText("X"));

  // expect todo to be gone
  expect(list.queryByText("write tests")).not.toBeInTheDocument();
});

it("can mark todo as complete", () => {
    const list = render(<TodoList />);
    addTodo(list);

    fireEvent.click(list.getByText("Mark as complete"));

    expect(list.queryByText("write tests")).toHaveStyle("textDecoration: line-through");

    fireEvent.click(list.getByText("Mark as complete"));

    expect(list.queryByText("write tests")).not.toHaveStyle("textDecoration: line-through");
})
