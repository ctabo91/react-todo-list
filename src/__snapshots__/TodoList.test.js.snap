// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="TodoList"
  >
    <div
      class="NewTodoForm"
    >
      <form>
        <div>
          <label
            for="task"
          >
            Task:
          </label>
          <input
            id="task"
            name="task"
            type="text"
            value=""
          />
        </div>
        <button
          id="newTodoButton"
        >
          Add Todo!
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
